Address privilege dropping

aleixpol opened the pull request:
Our sddm-helper needs to stay a root process as it's in charge of dealing with PAM.  Alternatively the pam calls in PamBackend will fail.

To do so, we get helper processes to run xorg+greeter for x11-user and wayland-compositor+greeter for wayland. Obviously for general x11 don't need a helper as it runs as root.

davidedmundson committed:
Revert "Move utmp logic to Display"

This reverts commit 4c99b784a99db8028ca438edb1f4ddb03b3b6e8c.

davidedmundson committed:
add asserts

davidedmundson committed:
Teardown on exit

davidedmundson committed:
revert QProcess inheritance

aleixpol committed:
Move launching X11-User and Wayland into a separate process

aleixpol committed:
XOrgUserHelper: Initialise environment

aleixpol committed:
Make sure helpers have debug handlers

Or otherwise all information gets lost.

aleixpol committed:
x11user: Quit helper with the session

aleixpol committed:
x11 user: we also need to launch the server on non-greeter sessions

aleixpol committed:
UserSession::stop: Improve how we stop processes

* Wait a bit longer for non-greeter sessions in case they need to wrap
the session up
* Wait again after killing

aleixpol committed:
Stop the session process as soon as we know it's getting closed

aleixpol committed:
When not on rootful X, we don't need a new VT every time

aleixpol committed:
Only create a new VT if we really need it

davidedmundson commented:
It's a definite major regression that needs fixing before release. 
My patch to drop priveleges early is flawed, I found a few modules that need pam_close_session to be as root, and it's in the documentation.

Direction wise, I don't think we have a choice, so ++

Review wise, I think we can squash a lot of these commits.

davidedmundson commented on src/helper/HelperApp.cpp:
```diff
@@ -156,21 +189,29 @@ namespace SDDM {
+            if (env.value(QStringLiteral("XDG_SESSION_CLASS")) != QLatin1String("greeter")) {
+                // cache pid for session end
+                m_session->setCachedProcessId(m_session->processId());
+                utmpLogin(vt, displayId, m_user, m_session->processId(), true);
```
why do we call m_session->setCachedProcessId() from here rather than in
`Session::start()`

Then it's all encapsulated.

aleixpol replied:
I'll move it, but this also is coming from the reverted commit.

davidedmundson commented on src/helper/HelperApp.cpp:
```diff
@@ -156,21 +189,29 @@ namespace SDDM {
+            sessionOpened(true);
+
+            // write successful login to utmp/wtmp
+            QProcessEnvironment env = m_session->processEnvironment();
```
const 

on most these args

aleixpol replied:
This is just part of the review because of the reverted commit. Will address it because it's an easy thing to do.

davidedmundson commented on src/helper/xorguserhelper.cpp:
```diff
@@ -95,10 +105,14 @@ bool XOrgUserHelper::startProcess(const QString &cmd,
-    process->setProcessChannelMode(QProcess::ForwardedChannels);
     process->setProcessEnvironment(env);
-
+    process->setInputChannelMode(QProcess::ForwardedInputChannel);
```
why do we do this?

aleixpol replied:
As the comment above says:

    // Make sure to forward the input of this process into the Xorg
    // server, otherwise it will complain that only console users are allowed

davidedmundson commented on src/helper/HelperStartWayland.cpp:
```diff
@@ -0,0 +1,59 @@
+    QCoreApplication app(argc, argv);
+    if (argc != 3) {
+        QTextStream(stderr) << "Wrong number of arguments\n";
+        return 33;
```
I assume this was just to help with debugging.

I think we can go  to just return  -1

aleixpol replied:
Changed it to return 1 and 2.

davidedmundson commented on src/helper/HelperStartWayland.cpp:
```diff
@@ -0,0 +1,59 @@
+    });
+    QObject::connect(&helper, &WaylandHelper::failed, &app, [&app] {
+        QTextStream(stderr) << "Failed to start wayland session" << Qt::endl;
+        app.exit(35);
```
and here

aleixpol committed:
Make some local variables const

aleixpol committed:
Let UserSession cache its processId on its own

aleixpol commented:
@davidedmundson addressed your comments, I didn't fold the changes into where they were added for the cases where it was coming from the reverted code.

yan12125 commented:
This PR seems to break login for rootless X11. I got the following error from `journalctl`:

> Failed to take control of /dev/tty7: Operation not permitted

aleixpol commented:
> This PR seems to break login for rootless X11

Would you be able to create an issue and provide how you have it configured so I can reproduce and address it?

yan12125 commented:
Ah, someone beats me :) https://github.com/sddm/sddm/issues/1456
